### PR TITLE
MSEARCH-781: add bibframe search functionality

### PR DESCRIPTION
**Pull request by aleksei-pronichev**
### Purpose
_Implement bibframe indexing in mod-search_

### Approach
_Implement bibframe indexing in mod-search_

### Changes Checklist
- [x] **API Changes**: Document any API paths, methods, request or response bodies changed, added, or removed.
- [ ] **Database Schema Changes**: Indicate any database schema changes and their impact. Confirm that migration scripts were created.
- [ ] **Interface Version Changes**: Indicate any changes to interface versions.
- [ ] **Interface Dependencies**: Document added or removed dependencies.
- [ ] **Permissions**: Document any changes to permissions.
- [ ] **Logging**: Confirm that logging is appropriately handled.
- [x] **Unit Testing**: Confirm that changed classes were covered by unit tests.
- [x] **Integration Testing**: Confirm that changed logic was covered by integration tests.
- [ ] **Manual Testing**: Confirm that changes were tested on local or dev environment.
- [x] **NEWS**: Confirm that the NEWS file is updated with relevant information about the changes made in this pull request.

### Related Issues
[MSEARCH-781](https://folio-org.atlassian.net/browse/MSEARCH-781)

### CODE REVIEW

File: src/main/java/org/folio/search/service/setter/bibframe/LccnNormalizer.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+import jakarta.validation.constraints.NotNull;
+import java.util.Optional;
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Service;
+
+@Log4j2
+@Service
+public class LccnNormalizer {
```
Comment (psmagin) [inline src/main/java/org/folio/search/service/setter/bibframe/LccnNormalizer.java:11]: As this LCCN normalizer is used only by bibframe and in the module exists another logic related to LCCN normalization I suggest combining all related to LCCN logic in one place as having it separated may confuse.

File: src/test/java/org/folio/search/service/setter/bibframe/LccnNormalizerTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    assertThat(lccnNormalizer.normalizeLccn("2020-2")).isEqualTo(Optional.of("2020000002"));
+  }
+
+  @Test
+  void shouldNormalizeSpacesAndHyphenAndForwardSlash() {
+    assertThat(lccnNormalizer.normalizeLccn("  20 20-2 //r23/AC")).isEqualTo(Optional.of("2020000002"));
+  }
+
+  @Test
+  void shouldReturnEmptyOptionalWhenLccnIsNotValid() {
```
Comment (psmagin) [inline src/test/java/org/folio/search/service/setter/bibframe/LccnNormalizerTest.java:44]: I suggest converting this test to a parametrized test.

File: src/main/java/org/folio/search/service/setter/bibframe/LccnNormalizer.java (hunk 1 above)
Comment (pkjacob) [reply src/main/java/org/folio/search/service/setter/bibframe/LccnNormalizer.java:11]: Hi @psmagin, @aleksei-pronichev 

This LccnNormalizer class is for normalizing Structure B LCCNs (Items acquired after 2001). I think the [normalizeLccn](https://github.com/folio-org/mod-search/blob/master/src/main/java/org/folio/search/utils/SearchUtils.java#L337) function in `SearchUtils` class is a generic normalizer that could be used for any kinds of LCCNs.

So, how about doing as follows?

Create an abstract class `LccnNormalizer` with two sub classes
1. `GenericLCCNNormalizer` -> Move the logic from [normalizeLccn](https://github.com/folio-org/mod-search/blob/master/src/main/java/org/folio/search/utils/SearchUtils.java#L337) function  to this class.
3. `StructureBLCCNormalizer` -> Rename this class to `StructureBLCCNormalizer`

Use `GenericLCCNNormalizer` in `AbstractLccnProcessor`
Use `StructureBLCCNormalizer` in `BibframeLccnProcessor`

Note: Documentation of StructureA and StructureB LCCNs - https://www.loc.gov/marc/lccn_structure.html#:~:text=On%20January%201%2C%202001%2C%20a,to%20make%20a%20LCCN%20unique.

File: src/test/java/org/folio/search/utils/SearchUtilsTest.java [hunk 3]
```diff
@@ -91,14 +90,6 @@ void getTotalPages_parameterized(long total, long expected) {
     assertThat(totalPages).isEqualTo(expected);
   }
 
-  @DisplayName("LCCN value normalization")
```
Comment (pkjacob) [inline src/test/java/org/folio/search/utils/SearchUtilsTest.java:94]: Moved this logic to DefaultLcccnNormalizer

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -106,6 +106,9 @@ folio:
       - name: search.consortium.instance
         numPartitions: ${KAFKA_CONSORTIUM_INSTANCE_TOPIC_PARTITIONS:50}
         replicationFactor: ${KAFKA_CONSORTIUM_INSTANCE_TOPIC_REPLICATION_FACTOR:}
+      - name: search.bibframe
+        numPartitions: ${KAFKA_BIBFRAME_TOPIC_PARTITIONS:3}
+        replicationFactor: ${KAFKA_BIBFRAME_TOPIC_REPLICATION_FACTOR:}
```
[the remark this answers is not in the document]
Comment (pkjacob) [reply src/main/resources/application.yml]: Done. Removed this line from the yaml file.

File: src/main/resources/application.yml [hunk 5]
```diff
@@ -135,6 +138,10 @@ folio:
         concurrency: ${KAFKA_LOCATION_CONCURRENCY:1}
         topic-pattern: (${folio.environment}\.)(.*\.)inventory\.location
         group-id: ${folio.environment}-mod-search-location-type-group
+      bibframe:
+        concurrency: ${KAFKA_BIBFRAME_CONCURRENCY:1}
```
[the remark this answers is not in the document]
Comment (pkjacob) [reply src/main/resources/application.yml:139]: Done. Updated README

File: src/main/java/org/folio/search/service/setter/bibframe/LccnNormalizer.java (hunk 1 above)
Comment (pkjacob) [reply src/main/java/org/folio/search/service/setter/bibframe/LccnNormalizer.java]: Done. Created new Classes and interfaces.
src/main/java/org/folio/search/service/lccn/LccnNormalizerStructureB.java
src/main/java/org/folio/search/service/lccn/LccnNormalizerStructureB.java
src/main/java/org/folio/search/service/lccn/DefaultLccnNormalizer.java

File: src/test/java/org/folio/search/service/setter/bibframe/LccnNormalizerTest.java (hunk 2 above)
Comment (pkjacob) [reply src/test/java/org/folio/search/service/setter/bibframe/LccnNormalizerTest.java]: Done

**Review by psmagin — APPROVED** [review]
Looks good!

Before merging please:
- fix Sonar issue
- create a ticket in JIRA mod-search project that will represent the work of the PR and add it in PR description and use in merge commit message
- update NEWS file

Use squash as a method for merging the PR and please use same commit style as the repository has: https://folio-org.atlassian.net/wiki/spaces/FOLIJET/pages/1400654/Conventional+Commit+Initiative

File: src/main/java/org/folio/search/utils/SearchUtils.java [hunk 6]
```diff
@@ -328,20 +329,6 @@ public static int getNumberOfRequests(Map<String, List<SearchDocumentBody>> requ
       .sum();
   }
 
-  /**
```
Comment (pkjacob) [inline src/main/java/org/folio/search/utils/SearchUtils.java:331]: Moved this logic to DefaultLccnNormalizer bean.

**Review by mukhiddin-yusuf — APPROVED** [review]

**Review by viacheslavpoliakov — APPROVED** [review]

**Review by PBobylev — APPROVED** [review]

**Pull request merged**